src/text/line.rs

```diff
@@ -586,31 +553,55 @@ impl WidgetRef for Line<'_> {
     fn render_ref(&self, area: Rect, buf: &mut Buffer) {
         let area = area.intersection(buf.area);
         buf.set_style(area, self.style);
-        let width = self.width() as u16;
-        let mut x = area.left();
-        let line = if width > area.width {
-            self.truncated(area.width)
+        let line_width = self.width() as u16;
+        if line_width <= area.width {
+            let offset = match self.alignment {
+                Some(Alignment::Center) => (area.width.saturating_sub(line_width)) / 2,
+                Some(Alignment::Right) => area.width.saturating_sub(line_width),
+                Some(Alignment::Left) | None => 0,
+            };
+            render_spans(&self.spans, area, buf, offset, 0);
         } else {
             let offset = match self.alignment {
-                Some(Alignment::Center) => (area.width.saturating_sub(width)) / 2,
-                Some(Alignment::Right) => area.width.saturating_sub(width),
+                Some(Alignment::Center) => (line_width.saturating_sub(area.width)) / 2,
+                Some(Alignment::Right) => line_width.saturating_sub(area.width),
                 Some(Alignment::Left) | None => 0,
             };
-            x = x.saturating_add(offset);
-            self.to_owned()
+            render_spans(&self.spans, area, buf, 0, offset);
         };
-        for span in &line.spans {
-            let span_width = span.width() as u16;
-            let span_area = Rect {
-                x,
-                width: span_width.min(area.right().saturating_sub(x)),
-                ..area
-            };
-            span.render(span_area, buf);
-            x = x.saturating_add(span_width);
-            if x >= area.right() {
-                break;
-            }
+    }
+}
+
+fn render_spans(
+    spans: &[Span],
+    area: Rect,
+    buf: &mut Buffer,
+    mut x_offset: u16,
+    mut span_offset: u16,
+) {
+    for span in spans {
+        let span_width = span.width() as u16;
+        // ignore spans that are completely before the offset
+        if span_offset >= span_width {
+            span_offset -= span_width;
+            continue;
+        }
+        let area = Rect {
+            x: area.x.saturating_add(x_offset),
+            width: area.width.saturating_sub(x_offset),
+            ..area
+        };
+        let is_span_partially_visible = span_offset > 0;
+        if is_span_partially_visible {
+            // only render the right side of the partially visible span
+            let (_left, right) = span.split_at_width(span_offset as usize);
```

Comment:
I don't quite understand why we only render the right side when a `Span` is partially visible?

Reply:
vec![Span::raw("Hello"), Span::raw("World")]
width 4, alignment centered -> "lo" (right side of the first span that is past the offset) then "Wo"
width 7, alignment right -> same